Mathlib/CategoryTheory/Bicategory/NaturalTransformation/Pseudo.lean

```diff
@@ -0,0 +1,227 @@
+/-
+Copyright (c) 2024 Calle Sönne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Calle Sönne
+-/
+
+import Mathlib.CategoryTheory.Bicategory.Functor.Pseudofunctor
+import Mathlib.CategoryTheory.Bicategory.NaturalTransformation.Oplax
+
+/-!
+
+# Strong transformations of pseudofunctors
+
+There are three types of transformations between pseudofunctors, depending on the direction
+or invertibility of the 2-morphism witnessing the naturality condition.
+
+In this file we define strong transformations, which require the 2-morphism to be invertible.
+
+## Main definitions
+
+* `StrongTrans F G` : strong transformations between pseudofunctors `F` and `G`.
+* `mkOfOplax η η'` : given an oplax transformation `η` such that each component 2-morphism
+  is an isomorphism, `mkOfOplax` gives the corresponding strong transformation.
+* `StrongTrans.vcomp η θ` : the vertical composition of strong transformations `η` and `θ`.
+
+Using this we obtain a `CategoryStruct` on pseudofunctors, where the arrows are given by
+strong transformations. See `Pseudofunctor.categoryStruct`.
+
+## References
+* [Niles Johnson, Donald Yau, *2-Dimensional Categories*](https://arxiv.org/abs/2002.06055)
+
+-/
+
+namespace CategoryTheory.Pseudofunctor
+
+open Category Bicategory Oplax
+
+open scoped Bicategory
+
+universe w₁ w₂ v₁ v₂ u₁ u₂
+
+variable {B : Type u₁} [Bicategory.{w₁, v₁} B] {C : Type u₂} [Bicategory.{w₂, v₂} C]
+
+/-- A strong transformation between pseudofunctors `F` and `G` is a natural transformation
+that is "natural up to 2-isomorphisms".
+More precisely, it consists of the following:
+* a 1-morphism `η.app a : F.obj a ⟶ G.obj a` for each object `a : B`.
+* a 2-isomorphism `η.naturality f : F.map f ≫ app b ≅ app a ≫ G.map f` for each 1-morphism
+`f : a ⟶ b`.
+* These 2-isomorphisms satisfy the naturality condition, and preserve the identities and the
+compositions modulo some adjustments of domains and codomains of 2-morphisms.
+-/
+structure StrongTrans (F G : Pseudofunctor B C) where
+  /-- The component 1-morphisms of a strong transformation. -/
+  app (a : B) : F.obj a ⟶ G.obj a
+  /-- The 2-isomorphisms underlying the strong naturality constraint. -/
+  naturality {a b : B} (f : a ⟶ b) : F.map f ≫ app b ≅ app a ≫ G.map f
+  /-- Naturality of the strong naturality constraint. -/
+  naturality_naturality :
+    ∀ {a b : B} {f g : a ⟶ b} (η : f ⟶ g),
+      F.map₂ η ▷ app b ≫ (naturality g).hom = (naturality f).hom ≫ app a ◁ G.map₂ η := by
```

Comment:
A bit cleaner this way IMO
```suggestion
  naturality_naturality {a b : B} {f g : a ⟶ b} (η : f ⟶ g) :
    F.map₂ η ▷ app b ≫ (naturality g).hom = (naturality f).hom ≫ app a ◁ G.map₂ η := by
```
same for oplax functors

Reply:
Done!